### ISSUE TITLE
Should response/request.header[T] be ModeledCustomHeader aware?

### ISSUE THREAD

**Issue by ktoso** [akka/akka-http#1102]
This requires a bunch of hacks though, as the impl that makes it work in akka-http is the `HeaderMagnet` infra which we do not have immediately available in akka-http-core, and I'm not sure if we want to have it there.

https://github.com/akka/akka-http/blob/master/akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala#L84

We can make it work using some smarts however making it possible for Java should likely use the partial function style as we suggest anyway for the routing DSL for it.

**Comment by czare1** [comment]
At least the documentation should mention that it won't work this way.

**Comment by jroper** [comment]
Agreed, the documentation does need to be updated, there's no mention [here](https://doc.akka.io/docs/akka-http/10.0/common/http-model.html) that if you create a custom header, and try to use it with `request.header[MyCustomHeader]`, it won't work.

**Comment by jroper** [comment]
I think we need to do more than just document this. Even we are [tripped up](https://github.com/akka/akka-grpc/issues/1897) by this, I know I've spent multiple hours staring at my code not understanding why it doesn't work because of this. If that's our experience, what of a regular akka-http user?

**Comment by johanandren** [comment]
If we implement this, it still means users must remember to register modeled header types, if they forget, we'd still have the same problem as now where it seems correct but always returns `None`.
